### ISSUE TITLE
Issue with Huggingface accelerate + FSDP,  No model update unless prepare is  done together for model, optimizer, lr_schedular

### ISSUE THREAD

**Issue by amarazad** [pytorch/pytorch#107843]
### 🐛 Describe the bug

I am using HF accelerate with FSDP. HF recommends to first do model prepare (accelerate.prepare()) and then do optimizer prepare. But when I use that way it does not update the model  parameters, i.e. model does not train. While debugging that model sharding has happen, gradient is also calculated. 
Here is code snippet (setting 1): 
```
model = accelerator.prepare(model) 
train_dataloader, valid_dataloader, optimizer, lr_scheduler = accelerator.prepare( train_dataloader, valid_dataloader, optimizer, lr_scheduler) 
```

However,  the parameter update works, i.e. model training happens fine if I  prepare together.   Though the code is working but i get the warning message 
`WARNING - FSDP Warning: When using FSDP, it is efficient and recommended to call prepare for the model before creating the optimizer`

This setting is doing model parameters update (working version - setting 2): 
    `model, train_dataloader, valid_dataloader, optimizer, lr_scheduler = accelerator.prepare( model, train_dataloader, valid_dataloader, optimizer, lr_scheduler) `

Initialization of acclerate is same for both, as follows:
```
from accelerate import Accelerator
from torch.distributed.fsdp import (
        FullyShardedDataParallel as FSDP,
        StateDictType,
        FullStateDictConfig
        )
from accelerate.utils import InitProcessGroupKwargs

accelerator_log_kwargs = {
            'kwargs_handlers':[InitProcessGroupKwargs(timeout=1800)],
            'gradient_accumulation_steps': 1,
            }
    accelerator_log_kwargs["log_with"] = "tensorboard"
    accelerator_log_kwargs["project_dir"] = dirs["logs_dir"]

    ## Initialize the accelerator
    accelerator = (
            Accelerator(**accelerator_log_kwargs)
            )
```
   
The accelerate config file for FSDP is 
```
compute_environment: LOCAL_MACHINE
distributed_type: FSDP
downcast_bf16: 'bf16'
fsdp_config:
  fsdp_auto_wrap_policy: TRANSFORMER_BASED_WRAP
  fsdp_backward_prefetch_policy: BACKWARD_PRE
  fsdp_offload_params: false
  fsdp_sharding_strategy: 1
  fsdp_state_dict_type: FULL_STATE_DICT
  fsdp_transformer_layer_cls_to_wrap: CodeGenBlock
machine_rank: 0
main_training_function: main
mixed_precision: 'bf16'
num_machines: 1
num_processes: 2
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false
```

### Versions

```
PyTorch version: 2.0.1
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Red Hat Enterprise Linux release 8.8 (Ootpa) (x86_64)
GCC version: (GCC) 8.5.0 20210514 (Red Hat 8.5.0-18)
Clang version: Could not collect
CMake version: version 3.20.2
Libc version: glibc-2.28

Python version: 3.11.3 | packaged by conda-forge | (main, Apr  6 2023, 08:57:19) [GCC 11.3.0] (64-bit runtime)
Python platform: Linux-4.18.0-477.15.1.el8_8.x86_64-x86_64-with-glibc2.28
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA A100-SXM4-80GB
GPU 1: NVIDIA A100-SXM4-80GB

Nvidia driver version: 535.54.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True


Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] torch==2.0.1
[pip3] torchaudio==2.0.2
[pip3] torchinfo==1.8.0
[pip3] torchvision==0.15.2
[pip3] triton==2.0.0
[conda] blas                      2.116                       mkl    conda-forge
[conda] blas-devel                3.9.0            16_linux64_mkl    conda-forge
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] libblas                   3.9.0            16_linux64_mkl    conda-forge
[conda] libcblas                  3.9.0            16_linux64_mkl    conda-forge
[conda] liblapack                 3.9.0            16_linux64_mkl    conda-forge
[conda] liblapacke                3.9.0            16_linux64_mkl    conda-forge
[conda] mkl                       2022.1.0           h84fe81f_915    conda-forge
[conda] mkl-devel                 2022.1.0           ha770c72_916    conda-forge
[conda] mkl-include               2022.1.0           h84fe81f_915    conda-forge
[conda] numpy                     1.24.3          py311h64a7726_0    conda-forge
[conda] pytorch                   2.0.1           py3.11_cuda11.7_cudnn8.5.0_0    pytorch
[conda] pytorch-cuda              11.7                 h778d358_5    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                2.0.2               py311_cu117    pytorch
[conda] torchinfo                 1.8.0                    pypi_0    pypi
[conda] torchtriton               2.0.0                     py311    pytorch
[conda] torchvision               0.15.2              py311_cu117    pytorch
```

Other relevant libraries version
```
transformers
Version: 4.31.0
accelerate
Version: 0.21.0
```

I have tried that with other version of python (3.9) and also other version of accelerate (0.18, 0.19) and transformers (4.24.0), i find the same (model doesn't train with setting 1). 
When I change the accelerate config file for no fsdp, the setting 1 works fine.  The accelerate config without FSDP in which setting 1 does training is.
```
compute_environment: LOCAL_MACHINE
distributed_type: MULTI_GPU
downcast_bf16: 'no'
gpu_ids: all
machine_rank: 0
main_training_function: main
mixed_precision: 'no'
num_machines: 1
num_processes: 2
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false
```

Can some one help to fix this ?

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu

**Comment by awgu** [comment]
This issue seems better for the HuggingFace accelerate repo?

**Comment by amarazad** [comment]
@awgu Does this issue needs to be transferred to Huggingface accelerate ? Please let me know.

**Comment by awgu** [comment]
Yes. Please transfer it.
